### PR TITLE
Update BouncyCastle-JCA ruleset

### PR DESCRIPTION
**Pull request by enriozuni**
JavaCryptographicArchitecture and BouncyCastle-JCA rulesets should mostly have the same structure and rules since the latter ruleset describes an implementation of the JCA's APIs by the BouncyCastle third-party provider. The only difference between the two rulesets is the CONSTRAINTS section, where they might be providing the implementation of the JCA's APIs through different cryptographic algorithms.

The pull requests [1](https://github.com/CROSSINGTUD/Crypto-API-Rules/pull/68) and [2](https://github.com/CROSSINGTUD/Crypto-API-Rules/pull/67) introduce many changes in the JavaCryptographicArchitecture ruleset. These changes were not included in the BouncyCastle-JCA ruleset.

In order to keep consistency between the two rulesets, this PR updates and refactors all the BouncyCastle-JCA ruleset. Mainly, the ORDER, REQUIRES, and ENSURES section of the rules are updated. The version of BouncyCastle-JCA ruleset is also increased from 0.3 to 0.4 since this PR contains a lot of changes.

### CODE REVIEW

File: BouncyCastle-JCA/src/KeyPair.crysl [hunk 1]
```diff
@@ -1,26 +1,24 @@
...
-
-	pu : retPub = getPublic();
-	pr : retPriv = getPrivate();
-	Gets := pu | pr;
+	con: KeyPair(consPub, consPriv);
+	pu: retPub = getPublic();
+	pr: retPriv = getPrivate();
 
 ORDER
-	co?, (pu*, pr*)*
```
Comment (schlichtig) [inline BouncyCastle-JCA/src/KeyPair.crysl:17]: Did you remove the '?' on purpose?

File: BouncyCastle-JCA/src/KeyStoreBuilderParameters.crysl [hunk 2]
```diff
@@ -4,10 +4,10 @@ OBJECTS
 	java.security.KeyStore.Builder builder;
 	
 EVENTS
-	C: KeyStoreBuilderParameters(builder);
+	c: KeyStoreBuilderParameters(builder);
```
Comment (schlichtig) [inline BouncyCastle-JCA/src/KeyStoreBuilderParameters.crysl:7]: Is this a constructor? Would namig it 'con' make sense for consistency with your other renaming?

File: BouncyCastle-JCA/src/SSLContext.crysl [hunk 3]
```diff
@@ -28,8 +28,8 @@ CONSTRAINTS
 	protocol in {"TLSv1.2"};
 
 REQUIRES
-	generatedKeyManager[kms];
-	generatedTrustManager[tms];
+	generatedKeyManagers[kms];
+	generatedTrustManagers[tms];
```
Comment (schlichtig) [inline BouncyCastle-JCA/src/SSLContext.crysl:32]: Why di you add the 's'?

File: BouncyCastle-JCA/src/KeyFactory.crysl [hunk 4]
```diff
@@ -12,14 +14,17 @@ EVENTS
     gPriv: generatePrivate(keySpec);
     gPubl: generatePublic(keySpec);
     
-    Gens := gPriv | gPubl;
-
 ORDER
-    Gets, Gens
+    Gets, (gPriv* | gPubl*)*
```
Comment (schlichtig) [inline BouncyCastle-JCA/src/KeyFactory.crysl:18]: Why did you change this?

File: BouncyCastle-JCA/src/CertPathTrustManagerParameters.crysl [hunk 5]
```diff
@@ -4,10 +4,10 @@ OBJECTS
 	java.security.cert.CertPathParameters params;
 
 EVENTS
-	C: CertPathTrustManagerParameters(params);	
+	c: CertPathTrustManagerParameters(params);	
 	
 ORDER
-	C
+	c
```
Comment (schlichtig) [inline BouncyCastle-JCA/src/CertPathTrustManagerParameters.crysl:10]: Can you check for consistency in the naming? You changed 'cP' to clear and often 'c1' to 'con'. If 'c' here is a constructor, shouldn't it be 'con'?

File: BouncyCastle-JCA/src/Cipher.crysl [hunk 6]
```diff
@@ -120,7 +123,8 @@ REQUIRES
 	mode(transformation) in {"GCM"} => preparedGCM[params];
 	
 ENSURES
-	encrypted[pre_ciphertext, pre_plaintext] after updates; 
+	generatedCipher[this] after Inits;
+	encrypted[pre_ciphertext, pre_plaintext] after Updates; 
 	encrypted[cipherText, plainText];
 	encrypted[cipherBuffer, plainBuffer];
 	wrappedKey[wrappedKeyBytes, wrappedKey];
```
Comment (schlichtig) [inline BouncyCastle-JCA/src/Cipher.crysl:130]: If you make changes, can you add the empty line at the end of the rules? (optionaly)

File: BouncyCastle-JCA/src/KeyPair.crysl (hunk 1 above)
Comment (enriozuni) [reply BouncyCastle-JCA/src/KeyPair.crysl:17]: Yes, because it was a typo

File: BouncyCastle-JCA/src/KeyStoreBuilderParameters.crysl (hunk 2 above)
Comment (enriozuni) [reply BouncyCastle-JCA/src/KeyStoreBuilderParameters.crysl]: True. Thanks for the suggestion. I have already accommodated the required changes.

File: BouncyCastle-JCA/src/SSLContext.crysl (hunk 3 above)
Comment (enriozuni) [reply BouncyCastle-JCA/src/SSLContext.crysl:32]: Because there are two possible predicates that can be ensured in the respective [TrustManagerFactory](https://github.com/CROSSINGTUD/Crypto-API-Rules/blob/master/BouncyCastle-JCA/src/TrustManagerFactory.crysl#L32) and [KeyManagerFactory](https://github.com/CROSSINGTUD/Crypto-API-Rules/blob/master/BouncyCastle-JCA/src/KeyManagerFactory.crysl#L33) rules and the SSLContext rule requires the second predicate that both these rules can ensure.

File: BouncyCastle-JCA/src/KeyFactory.crysl (hunk 4 above)
Comment (enriozuni) [reply BouncyCastle-JCA/src/KeyFactory.crysl]: Here I used the fix that was introduced in the JCA's [KeyFactory](https://github.com/CROSSINGTUD/Crypto-API-Rules/blob/master/JavaCryptographicArchitecture/src/KeyFactory.crysl#L17) rule and I am not sure why this was changed there. My guess would be that this was changed to ensure the last predicates in the [KeyFactory](https://github.com/CROSSINGTUD/Crypto-API-Rules/blob/master/JavaCryptographicArchitecture/src/KeyFactory.crysl#L27) rule.
